### PR TITLE
Adding support for exporting models with variable length input/output to ONNX

### PR DESCRIPTION
**Pull request by pk-g**
Proposal: https://gist.github.com/pk-g/cc45ff8c5891b5699bffd883a87f13ae?fbclid=IwAR17bRA7Fks4APoZRYiNa93UkLdoFCpRDuIYEx0lNVyPTyaDAShbEnytiQo

### CODE REVIEW

File: test/onnx/test_operators.py [hunk 1]
```diff
@@ -249,6 +249,28 @@ def test_conv(self):
...
         self.assertONNX(nn.Conv2d(16, 13, 3, bias=False), x)
 
+    def test_conv_variable_length(self):
+        x = torch.ones(5, 3, 6, 6, requires_grad=True)
+        model = torch.nn.Conv2d(3, 2, 3)
+        y = model(x)
+
+        dynamic_axes = {'input_1': [0, 2, 3], 'output_1': {0: 'output_1_variable_dim_0', 1: 'output_1_variable_dim_1'}}
+        model_proto_name = 'conv2d.onnx'
+        torch.onnx.export(model, x, model_proto_name, verbose=True, input_names=["input_1"], output_names=["output_1"], example_outputs=y, dynamic_axes=dynamic_axes)
```
Comment (houseroad) [inline test/onnx/test_operators.py]: Do you mean when we use dynamic_axes, we must set input_names and output_names as well?

Comment (pk-g) [reply test/onnx/test_operators.py]: Not exactly. User does NOT have to to provide `input_names`/ `output_names` on export API, if they already know what names are being used in the graph, they shall simply provide such names as keys to `dynamic_axes` dictionary and NOT specify `input_names` / `output_names` . however, if they do not know the names OR intend to use different names on exported model, they also have the option. They can do so by providing those names in  `input_names` / `output_names` and using them in `dynamic_axes`.

**Comment by pk-g** [comment]
@houseroad gentle reminder in case there are more feedback on this. thanks!

File: torch/onnx/utils.py [hunk 2]
```diff
@@ -115,6 +115,19 @@ def export(model, args, f, export_params=True, verbose=False, training=False,
         strip_doc_string (bool, default True): if True, strips the field
             "doc_string" from the exported model, which information about the stack
             trace.
+        example_outputs: example outputs of the model that is being exported.
+        dynamic_axes (dict<string, dict<int, string>> or dict<string, list(int)>, default empty dict): 
+            a dictionary to specify dynamic axes of input/output. The dictionary takes an input/output 
+            name as key and maps it one of the values below:
+            (1)A list of integers specifiying the dynamic axes of provided input. In this scenario
```
Comment (houseroad) [inline torch/onnx/utils.py]: Provide two examples here to help users to understand?

File: torch/onnx/utils.py [hunk 3]
```diff
@@ -128,7 +141,7 @@ def export(model, args, f, export_params=True, verbose=False, training=False,
     _export(model, args, f, export_params, verbose, training, input_names, output_names,
             operator_export_type=operator_export_type, opset_version=opset_version,
             _retain_param_name=_retain_param_name, do_constant_folding=do_constant_folding,
-            strip_doc_string=strip_doc_string)
+            strip_doc_string=strip_doc_string, example_outputs=example_outputs, dynamic_axes=dynamic_axes)
```
Comment (houseroad) [inline torch/onnx/utils.py]: please don't fix the example_outputs in diff, it's already fixed in another ongoing one.

File: torch/onnx/utils.py [hunk 4]
```diff
@@ -115,6 +115,19 @@ def export(model, args, f, export_params=True, verbose=False, training=False,
         strip_doc_string (bool, default True): if True, strips the field
             "doc_string" from the exported model, which information about the stack
             trace.
+        example_outputs: example outputs of the model that is being exported.
+        dynamic_axes (dict<string, dict<int, string>> or dict<string, list(int)>, default empty dict):
```
Comment (houseroad) [inline torch/onnx/utils.py]: what's the meaning of key and values?

**Review by houseroad — COMMENTED** [review]
I think we should ask users to specify the input_names/output_names if they use dynamic_axes. Otherwise, they have to guess the name of the inputs/outputs, because they don't have the control of them.

File: torch/onnx/utils.py (hunk 4 above)
Comment (pk-g) [reply torch/onnx/utils.py]: Sure, added.

File: torch/onnx/utils.py (hunk 3 above)
Comment (pk-g) [reply torch/onnx/utils.py]: Sure.

File: torch/onnx/utils.py (hunk 2 above)
Comment (pk-g) [reply torch/onnx/utils.py]: Sure, added.

**Comment by pk-g** [comment]
@houseroad during _validate_dynamic_axes method, we are generating a warning if names are not matching or not provided.

**Comment by pk-g** [comment]
@houseroad gentle reminder ... thanks!

**Review by houseroad — APPROVED** [review]
LGTM

**Pull request closed without merging**